src/IO/WriteBuffer.cpp

```diff
@@ -0,0 +1,33 @@
+#include "WriteBuffer.h"
+
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+/// Calling finalize() in the destructor of derived classes is a bad practice.
+/// This causes objects to be left on the remote FS when a write operation is rolled back.
+/// Do call finalize() explicitly, before this call you have no guarantee that the file has been written
+WriteBuffer::~WriteBuffer()
+{
+    // That destructor could be call with finalized=false in case of exceptions
+    if (count() > 0 && !finalized)
+    {
+        /// It is totally OK to destroy instance without finalization when an exception occurs
+        /// However it is suspicious to destroy instance without finalization at the green path
```

Comment:
Wow, that's a great idea about exceptions, I didn't think about it and tried to add try/catch in many places :)